### PR TITLE
perf(dbless) improve proxy long tail latency for DB-less reload [CT-321]

### PR DESCRIPTION
**Pull request by dndx**
Previously, when a large DB-less config is processed by Kong, Kong will
be CPU bound when parsing/validating the config and rebuilding the router.
This could cause undesirable proxy latency behavior during reloads.

This PR introduces periodic yielding points so that reload tasks no
longer hogs the CPU until reload completely finishes. Request processing
will be able to continue even while DB-less reload and router rebuild is
happening.

CT-321

### CODE REVIEW

File: kong/db/declarative/init.lua [hunk 1]
```diff
@@ -587,6 +590,8 @@ function declarative.load_into_cache(entities, meta, hash, shadow)
   local transform = meta._transform == nil and true or meta._transform
 
   for entity_name, items in pairs(entities) do
+    yield()
```
Comment (bungle) [inline kong/db/declarative/init.lua:593]:
```suggestion
    yield(true)
```

File: kong/db/declarative/init.lua [hunk 2]
```diff
@@ -786,6 +793,8 @@ function declarative.load_into_cache(entities, meta, hash, shadow)
   end
 
   for tag_name, tags in pairs(tags_by_name) do
+    yield()
```
Comment (bungle) [inline kong/db/declarative/init.lua:796]:
```suggestion
    yield(true)
```

File: kong/router.lua [hunk 3]
```diff
@@ -1418,12 +1422,16 @@ function _M.new(routes)
...
 
+  yield()
+
   -- the number of categories to iterate on for this instance of the router
   local categories_len = #categories_weight_sorted
 
   sort(prefix_uris, sort_uris)
 
   for _, category in pairs(categories) do
+    yield()
```
Comment (bungle) [inline kong/router.lua:1433]:
```suggestion
    yield(true)
```

**Review by bungle — CHANGES_REQUESTED** [review]

File: kong/tools/utils.lua [hunk 4]
```diff
@@ -1423,4 +1426,25 @@ local topological_sort do
...
+          ngx_sleep(0)
+        end
+
+        return
+      end
+
+      ngx_sleep(0)
+    end
+  end
+end
```
Comment (bungle) [inline kong/tools/utils.lua:1447]:
```suggestion
  function _M.yield(in_loop)
    local phase = get_phase()
    if phase == "init" or phase == "init_worker"  then
      return
    end
    if in_loop then
      counter = counter + 1
        if counter % YIELD_ITERATIONS ~= 0 then
          return
        end
        counter = 0
      end
      ngx_sleep(0)
    end
  end
end
```

**Review by bungle — CHANGES_REQUESTED** [review]

File: kong/db/declarative/init.lua (hunk 1 above)
Comment (dndx) [reply kong/db/declarative/init.lua:593]: Correct me if I'm wrong, but `for entity_name, items in pairs(entities)` only iterates known entities by entity name, it does not represents individual entity objects.

Comment (dndx) [reply kong/db/declarative/init.lua:593]: Same idea for the following ones, they all have small loops.

**Review by bungle — APPROVED** [review]

**Pull request merged**